src/jit/morph.cpp

```diff
@@ -7656,52 +7618,25 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
-    // not going to return and the helper doesn't have enough info to safely poll,
-    // so we poll before the tail call, if the block isn't already safe.  Since
-    // tail call via helper is a slow mechanism it doen't matter whether we emit
-    // GC poll.  This is done to be in parity with Jit64. Also this avoids GC info
-    // size increase if all most all methods are expected to be tail calls (e.g. F#).
-    //
-    // Note that we can avoid emitting GC-poll if we know that the current BB is
-    // dominated by a Gc-SafePoint block.  But we don't have dominator info at this
-    // point.  One option is to just add a place holder node for GC-poll (e.g. GT_GCPOLL)
-    // here and remove it in lowering if the block is dominated by a GC-SafePoint.  For
-    // now it not clear whether optimizing slow tail calls is worth the effort.  As a
-    // low cost check, we check whether the first and current basic blocks are
-    // GC-SafePoints.
-    //
-    // Fast Tail call as epilog+jmp - No need to insert GC-poll. Instead, fgSetBlockOrder()
-    // is going to mark the method as fully interruptible if the block containing this tail
-    // call is reachable without executing any call.
-    if (canFastTailCall || (fgFirstBB->bbFlags & BBF_GC_SAFE_POINT) || (compCurBB->bbFlags & BBF_GC_SAFE_POINT) ||
-        !fgCreateGCPoll(GCPOLL_INLINE, compCurBB))
-    {
-        // We didn't insert a poll block, so we need to morph the call now
-        // (Normally it will get morphed when we get to the split poll block)
-        GenTree* temp = fgMorphCall(call);
-        noway_assert(temp == call);
-    }
-
-    // Tail call via helper: we just call CORINFO_HELP_TAILCALL, and it jumps to
-    // the target. So we don't need an epilog - just like CORINFO_HELP_THROW.
-    //
-    // Fast tail call: in case of fast tail calls, we need a jmp epilog and
-    // hence mark it as BBJ_RETURN with BBF_JMP flag set.
-    noway_assert(compCurBB->bbJumpKind == BBJ_RETURN);
-
-    if (canFastTailCall)
+    // For helper-based tailcalls we transform into multiple calls where the
+    // last call is a fast tailcall. This will modify fgMorphStmt.
+    if (call->IsTailCallViaHelper())
     {
-        compCurBB->bbFlags |= BBF_HAS_JMP;
+        fgMorphTailCallViaHelper(call, tailCallHelp);
     }
     else
     {
-        compCurBB->bbJumpKind = BBJ_THROW;
+        fgMorphCall(call);
     }
 
-    // For non-void calls, we return a place holder which will be
-    // used by the parent GT_RETURN node of this call.
+    // Fast tail call: in case of fast tail calls, we need a jmp epilog and
+    // hence mark it as BBJ_RETURN with BBF_JMP flag set.
+    noway_assert(compCurBB->bbJumpKind == BBJ_RETURN);
+    compCurBB->bbFlags |= BBF_HAS_JMP;
 
+    // For non-void calls, we return a place holder which will be used by the
+    // parent GT_RETURN node of this call (note that we have already removed
+    // this return above).
```

Comment:
Right, this took me a while to understand too (this is not new), and I wrote the comment after to explain what is actually happening, but I will try to explain it more clearly in the comment.

The comment is actually talking about the return node "on the call stack". The return node has been removed from the tree, but morph of the call was still called from the function that morphs a `GT_RET` node, so when we return from morphing the call, that node is still being processed. I will make it clearer. It would also be nice to clean this up more but I am a little afraid of making changes in morph as I am not too familiar with all the bits and pieces and conventions.